File: lib/ecs_stack.py
"""
ecs_stack.py

ECS Fargate cluster with task definitions for healthcare analytics platform
Includes IAM roles and security configurations
"""

from typing import Optional, List
import json

import pulumi
from pulumi import ResourceOptions, Output
import pulumi_aws as aws


class ECSStackArgs:
    """Arguments for ECS stack"""

    def __init__(
        self,
        environment_suffix: str,
        tags: dict,
        vpc_id: Output[str],
        private_subnet_ids: List[Output[str]],
        redis_endpoint: Output[str],
        redis_port: Output[int],
        redis_secret_arn: Output[str]
    ):
        self.environment_suffix = environment_suffix
        self.tags = tags
        self.vpc_id = vpc_id
        self.private_subnet_ids = private_subnet_ids
        self.redis_endpoint = redis_endpoint
        self.redis_port = redis_port
        self.redis_secret_arn = redis_secret_arn


class ECSStack(pulumi.ComponentResource):
    """
    ECS Fargate cluster with task definitions
    Includes IAM roles for task execution and application permissions
    """

    def __init__(
        self,
        name: str,
        args: ECSStackArgs,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('tap:ecs:ECSStack', name, None, opts)

        # Create security group for ECS tasks
        self.ecs_task_sg = aws.ec2.SecurityGroup(
            f"ecs-task-sg-{args.environment_suffix}",
            vpc_id=args.vpc_id,
            description='Security group for ECS tasks',
            ingress=[
                aws.ec2.SecurityGroupIngressArgs(
                    protocol='tcp',
                    from_port=8080,
                    to_port=8080,
                    cidr_blocks=['10.0.0.0/16'],
                    description='Allow inbound traffic to application'
                )
            ],
            egress=[
                aws.ec2.SecurityGroupEgressArgs(
                    protocol='-1',
                    from_port=0,
                    to_port=0,
                    cidr_blocks=['0.0.0.0/0'],
                    description='Allow all outbound traffic'
                )
            ],
            tags={
                **args.tags,
                'Name': f'ecs-task-sg-{args.environment_suffix}'
            },
            opts=ResourceOptions(parent=self)
        )

        # Create ECS cluster
        self.cluster = aws.ecs.Cluster(
            f"ecs-cluster-{args.environment_suffix}",
            name=f"healthcare-analytics-{args.environment_suffix}",
            settings=[
                aws.ecs.ClusterSettingArgs(
                    name='containerInsights',
                    value='enabled'
                )
            ],
            tags={
                **args.tags,
                'Name': f'ecs-cluster-{args.environment_suffix}'
            },
            opts=ResourceOptions(parent=self)
        )

        # Create IAM role for ECS task execution
        self.task_execution_role = aws.iam.Role(
            f"ecs-task-execution-role-{args.environment_suffix}",
            name=f"ecs-task-execution-{args.environment_suffix}",
            assume_role_policy=json.dumps({
                'Version': '2012-10-17',
                'Statement': [{
                    'Effect': 'Allow',
                    'Principal': {
                        'Service': 'ecs-tasks.amazonaws.com'
                    },
                    'Action': 'sts:AssumeRole'
                }]
            }),
            tags={
                **args.tags,
                'Name': f'ecs-task-execution-role-{args.environment_suffix}'
            },
            opts=ResourceOptions(parent=self)
        )

        # Attach AWS managed policy for ECS task execution
        aws.iam.RolePolicyAttachment(
            f"ecs-task-execution-policy-{args.environment_suffix}",
            role=self.task_execution_role.name,
            policy_arn='arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy',
            opts=ResourceOptions(parent=self)
        )

        # Add policy for Secrets Manager access
        self.secrets_policy = aws.iam.RolePolicy(
            f"ecs-secrets-policy-{args.environment_suffix}",
            role=self.task_execution_role.id,
            policy=args.redis_secret_arn.apply(
                lambda arn: json.dumps({
                    'Version': '2012-10-17',
                    'Statement': [{
                        'Effect': 'Allow',
                        'Action': [
                            'secretsmanager:GetSecretValue'
                        ],
                        'Resource': [arn]
                    }]
                })
            ),
            opts=ResourceOptions(parent=self)
        )

        # Create IAM role for ECS task
        self.task_role = aws.iam.Role(
            f"ecs-task-role-{args.environment_suffix}",
            name=f"ecs-task-{args.environment_suffix}",
            assume_role_policy=json.dumps({
                'Version': '2012-10-17',
                'Statement': [{
                    'Effect': 'Allow',
                    'Principal': {
                        'Service': 'ecs-tasks.amazonaws.com'
                    },
                    'Action': 'sts:AssumeRole'
                }]
            }),
            tags={
                **args.tags,
                'Name': f'ecs-task-role-{args.environment_suffix}'
            },
            opts=ResourceOptions(parent=self)
        )

        # Add CloudWatch Logs policy for task role
        self.logs_policy = aws.iam.RolePolicy(
            f"ecs-logs-policy-{args.environment_suffix}",
            role=self.task_role.id,
            policy=json.dumps({
                'Version': '2012-10-17',
                'Statement': [{
                    'Effect': 'Allow',
                    'Action': [
                        'logs:CreateLogGroup',
                        'logs:CreateLogStream',
                        'logs:PutLogEvents'
                    ],
                    'Resource': 'arn:aws:logs:*:*:*'
                }]
            }),
            opts=ResourceOptions(parent=self)
        )

        # Create CloudWatch log group
        self.log_group = aws.cloudwatch.LogGroup(
            f"ecs-log-group-{args.environment_suffix}",
            name=f"/ecs/healthcare-analytics-{args.environment_suffix}",
            retention_in_days=7,
            tags={
                **args.tags,
                'Name': f'ecs-log-group-{args.environment_suffix}'
            },
            opts=ResourceOptions(parent=self)
        )

        # Create ECS task definition
        self.task_definition = aws.ecs.TaskDefinition(
            f"ecs-task-def-{args.environment_suffix}",
            family=f"healthcare-analytics-{args.environment_suffix}",
            network_mode='awsvpc',
            requires_compatibilities=['FARGATE'],
            cpu='256',
            memory='512',
            execution_role_arn=self.task_execution_role.arn,
            task_role_arn=self.task_role.arn,
            container_definitions=Output.all(
                args.redis_endpoint,
                args.redis_port,
                args.redis_secret_arn,
                self.log_group.name
            ).apply(
                lambda args_list: json.dumps([{
                    'name': 'healthcare-analytics-app',
                    'image': 'nginx:latest',
                    'essential': True,
                    'portMappings': [{
                        'containerPort': 8080,
                        'protocol': 'tcp'
                    }],
                    'environment': [
                        {
                            'name': 'REDIS_ENDPOINT',
                            'value': args_list[0]
                        },
                        {
                            'name': 'REDIS_PORT',
                            'value': str(args_list[1])
                        },
                        {
                            'name': 'ENVIRONMENT',
                            'value': args.environment_suffix
                        }
                    ],
                    'secrets': [{
                        'name': 'REDIS_AUTH_TOKEN',
                        'valueFrom': f"{args_list[2]}:auth_token::"
                    }],
                    'logConfiguration': {
                        'logDriver': 'awslogs',
                        'options': {
                            'awslogs-group': args_list[3],
                            'awslogs-region': 'eu-west-1',
                            'awslogs-stream-prefix': 'ecs'
                        }
                    }
                }])
            ),
            tags={
                **args.tags,
                'Name': f'ecs-task-def-{args.environment_suffix}'
            },
            opts=ResourceOptions(parent=self)
        )

        # Export properties
        self.cluster_name = self.cluster.name
        self.cluster_arn = self.cluster.arn
        self.task_definition_arn = self.task_definition.arn

        self.register_outputs({
            'cluster_name': self.cluster_name,
            'cluster_arn': self.cluster_arn,
            'task_definition_arn': self.task_definition_arn
        })

File: archive/pulumi-ts/Pr6224/lib/ecs-stack.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable quotes */
/* eslint-disable @typescript-eslint/quotes */
/* eslint-disable prettier/prettier */


/**
 * ecs-stack.ts
 *
 * Creates ECS cluster, ALB, ECS services, task definitions, and auto-scaling policies.
 */

import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface EcsStackArgs {
  environmentSuffix: string;
  tags?: pulumi.Input<{ [key: string]: string }>;
  vpcId: pulumi.Output<string>;
  publicSubnetIds: pulumi.Output<string>[];
  privateSubnetIds: pulumi.Output<string>[];
  apiEcrUrl: pulumi.Output<string>;
  workerEcrUrl: pulumi.Output<string>;
  schedulerEcrUrl: pulumi.Output<string>;
  dbSecretArn: pulumi.Output<string>;
  apiKeySecretArn: pulumi.Output<string>;
}

export class EcsStack extends pulumi.ComponentResource {
  public readonly albDnsName: pulumi.Output<string>;
  public readonly clusterName: pulumi.Output<string>;

  constructor(
    name: string,
    args: EcsStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:ecs:EcsStack', name, args, opts);

    const {
      environmentSuffix,
      tags,
      vpcId,
      publicSubnetIds,
      privateSubnetIds,
      apiEcrUrl,
      workerEcrUrl,
      schedulerEcrUrl,
      dbSecretArn,
      apiKeySecretArn,
    } = args;

    // Create ECS Cluster with Container Insights enabled
    const cluster = new aws.ecs.Cluster(
      `ecs-cluster-${environmentSuffix}`,
      {
        name: `ecs-cluster-${environmentSuffix}`,
        settings: [
          {
            name: 'containerInsights',
            value: 'enabled',
          },
        ],
        tags: {
          Name: `ecs-cluster-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create CloudWatch Log Groups
    const apiLogGroup = new aws.cloudwatch.LogGroup(
      `api-service-logs-${environmentSuffix}`,
      {
        name: `/ecs/api-service-${environmentSuffix}`,
        retentionInDays: 7,
        tags: {
          Name: `api-service-logs-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const workerLogGroup = new aws.cloudwatch.LogGroup(
      `worker-service-logs-${environmentSuffix}`,
      {
        name: `/ecs/worker-service-${environmentSuffix}`,
        retentionInDays: 7,
        tags: {
          Name: `worker-service-logs-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const schedulerLogGroup = new aws.cloudwatch.LogGroup(
      `scheduler-service-logs-${environmentSuffix}`,
      {
        name: `/ecs/scheduler-service-${environmentSuffix}`,
        retentionInDays: 7,
        tags: {
          Name: `scheduler-service-logs-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create IAM role for ECS task execution
    const taskExecutionRole = new aws.iam.Role(
      `ecs-task-execution-role-${environmentSuffix}`,
      {
        name: `ecs-task-execution-role-${environmentSuffix}`,
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: {
                Service: 'ecs-tasks.amazonaws.com',
              },
            },
          ],
        }),
        tags: {
          Name: `ecs-task-execution-role-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Attach AWS managed policy for ECS task execution
    new aws.iam.RolePolicyAttachment(
      `ecs-task-execution-policy-${environmentSuffix}`,
      {
        role: taskExecutionRole.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy',
      },
      { parent: this }
    );

    // Create custom policy for Secrets Manager access
    const secretsPolicy = new aws.iam.Policy(
      `ecs-secrets-policy-${environmentSuffix}`,
      {
        name: `ecs-secrets-policy-${environmentSuffix}`,
        policy: pulumi
          .all([dbSecretArn, apiKeySecretArn])
          .apply(([dbArn, apiArn]) =>
            JSON.stringify({
              Version: '2012-10-17',
              Statement: [
                {
                  Effect: 'Allow',
                  Action: [
                    'secretsmanager:GetSecretValue',
                    'secretsmanager:DescribeSecret',
                  ],
                  Resource: [dbArn, apiArn],
                },
              ],
            })
          ),
        tags: {
          Name: `ecs-secrets-policy-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    new aws.iam.RolePolicyAttachment(
      `ecs-secrets-policy-attach-${environmentSuffix}`,
      {
        role: taskExecutionRole.name,
        policyArn: secretsPolicy.arn,
      },
      { parent: this }
    );

    // Create IAM role for ECS tasks
    const taskRole = new aws.iam.Role(
      `ecs-task-role-${environmentSuffix}`,
      {
        name: `ecs-task-role-${environmentSuffix}`,
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: {
                Service: 'ecs-tasks.amazonaws.com',
              },
            },
          ],
        }),
        tags: {
          Name: `ecs-task-role-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create security group for ALB
    const albSecurityGroup = new aws.ec2.SecurityGroup(
      `alb-sg-${environmentSuffix}`,
      {
        name: `alb-sg-${environmentSuffix}`,
        description: 'Security group for Application Load Balancer',
        vpcId: vpcId,
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTP from anywhere',
          },
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTPS from anywhere',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          Name: `alb-sg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create security group for ECS tasks
    const ecsSecurityGroup = new aws.ec2.SecurityGroup(
      `ecs-tasks-sg-${environmentSuffix}`,
      {
        name: `ecs-tasks-sg-${environmentSuffix}`,
        description: 'Security group for ECS tasks',
        vpcId: vpcId,
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 8080,
            toPort: 8080,
            securityGroups: [albSecurityGroup.id],
            description: 'Allow traffic from ALB',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          Name: `ecs-tasks-sg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create Application Load Balancer
    const alb = new aws.lb.LoadBalancer(
      `alb-${environmentSuffix}`,
      {
        name: `alb-${environmentSuffix}`,
        internal: false,
        loadBalancerType: 'application',
        securityGroups: [albSecurityGroup.id],
        subnets: publicSubnetIds,
        enableDeletionProtection: false,
        tags: {
          Name: `alb-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create target groups for each service
    const apiTargetGroup = new aws.lb.TargetGroup(
      `api-tg-${environmentSuffix}`,
      {
        name: `api-tg-${environmentSuffix}`,
        port: 8080,
        protocol: 'HTTP',
        vpcId: vpcId,
        targetType: 'ip',
        healthCheck: {
          enabled: true,
          path: '/health',
          interval: 30,
          timeout: 5,
          healthyThreshold: 3,
          unhealthyThreshold: 3,
          matcher: '200',
        },
        deregistrationDelay: 30,
        tags: {
          Name: `api-tg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const workerTargetGroup = new aws.lb.TargetGroup(
      `worker-tg-${environmentSuffix}`,
      {
        name: `worker-tg-${environmentSuffix}`,
        port: 8080,
        protocol: 'HTTP',
        vpcId: vpcId,
        targetType: 'ip',
        healthCheck: {
          enabled: true,
          path: '/health',
          interval: 30,
          timeout: 5,
          healthyThreshold: 3,
          unhealthyThreshold: 3,
          matcher: '200',
        },
        deregistrationDelay: 30,
        tags: {
          Name: `worker-tg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const schedulerTargetGroup = new aws.lb.TargetGroup(
      `scheduler-tg-${environmentSuffix}`,
      {
        name: `scheduler-tg-${environmentSuffix}`,
        port: 8080,
        protocol: 'HTTP',
        vpcId: vpcId,
        targetType: 'ip',
        healthCheck: {
          enabled: true,
          path: '/health',
          interval: 30,
          timeout: 5,
          healthyThreshold: 3,
          unhealthyThreshold: 3,
          matcher: '200',
        },
        deregistrationDelay: 30,
        tags: {
          Name: `scheduler-tg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create ALB listener
    const listener = new aws.lb.Listener(
      `alb-listener-${environmentSuffix}`,
      {
        loadBalancerArn: alb.arn,
        port: 80,
        protocol: 'HTTP',
        defaultActions: [
          {
            type: 'fixed-response',
            fixedResponse: {
              contentType: 'text/plain',
              messageBody: 'Not Found',
              statusCode: '404',
            },
          },
        ],
        tags: {
          Name: `alb-listener-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create listener rules for path-based routing
    new aws.lb.ListenerRule(
      `api-rule-${environmentSuffix}`,
      {
        listenerArn: listener.arn,
        priority: 100,
        conditions: [
          {
            pathPattern: {
              values: ['/api/*'],
            },
          },
        ],
        actions: [
          {
            type: 'forward',
            targetGroupArn: apiTargetGroup.arn,
          },
        ],
        tags: {
          Name: `api-rule-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    new aws.lb.ListenerRule(
      `worker-rule-${environmentSuffix}`,
      {
        listenerArn: listener.arn,
        priority: 200,
        conditions: [
          {
            pathPattern: {
              values: ['/worker/*'],
            },
          },
        ],
        actions: [
          {
            type: 'forward',
            targetGroupArn: workerTargetGroup.arn,
          },
        ],
        tags: {
          Name: `worker-rule-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    new aws.lb.ListenerRule(
      `scheduler-rule-${environmentSuffix}`,
      {
        listenerArn: listener.arn,
        priority: 300,
        conditions: [
          {
            pathPattern: {
              values: ['/scheduler/*'],
            },
          },
        ],
        actions: [
          {
            type: 'forward',
            targetGroupArn: schedulerTargetGroup.arn,
          },
        ],
        tags: {
          Name: `scheduler-rule-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Get current AWS region
    const currentRegion = aws.getRegionOutput();

    // Create task definitions for each service
    // FIXED: Using pulumi.all to properly resolve all Output values including log group names and region
    const apiTaskDefinition = new aws.ecs.TaskDefinition(
      `api-task-${environmentSuffix}`,
      {
        family: `api-service-${environmentSuffix}`,
        networkMode: 'awsvpc',
        requiresCompatibilities: ['FARGATE'],
        cpu: '512',
        memory: '1024',
        executionRoleArn: taskExecutionRole.arn,
        taskRoleArn: taskRole.arn,
        containerDefinitions: pulumi
          .all([apiEcrUrl, dbSecretArn, apiKeySecretArn, apiLogGroup.name, currentRegion.name])
          .apply(([ecrUrl, dbArn, apiArn, logGroupName, region]) =>
            JSON.stringify([
              {
                name: 'api-service',
                image: `${ecrUrl}:latest`,
                cpu: 512,
                memory: 1024,
                essential: true,
                portMappings: [
                  {
                    containerPort: 8080,
                    protocol: 'tcp',
                  },
                ],
                environment: [
                  { name: 'SERVICE_NAME', value: 'api-service' },
                  { name: 'ENVIRONMENT', value: environmentSuffix },
                ],
                secrets: [
                  { name: 'DB_CREDENTIALS', valueFrom: dbArn },
                  { name: 'API_KEYS', valueFrom: apiArn },
                ],
                logConfiguration: {
                  logDriver: 'awslogs',
                  options: {
                    'awslogs-group': logGroupName,
                    'awslogs-region': region,
                    'awslogs-stream-prefix': 'ecs',
                  },
                },
              },
            ])
          ),
        tags: {
          Name: `api-task-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const workerTaskDefinition = new aws.ecs.TaskDefinition(
      `worker-task-${environmentSuffix}`,
      {
        family: `worker-service-${environmentSuffix}`,
        networkMode: 'awsvpc',
        requiresCompatibilities: ['FARGATE'],
        cpu: '512',
        memory: '1024',
        executionRoleArn: taskExecutionRole.arn,
        taskRoleArn: taskRole.arn,
        containerDefinitions: pulumi
          .all([workerEcrUrl, dbSecretArn, apiKeySecretArn, workerLogGroup.name, currentRegion.name])
          .apply(([ecrUrl, dbArn, apiArn, logGroupName, region]) =>
            JSON.stringify([
              {
                name: 'worker-service',
                image: `${ecrUrl}:latest`,
                cpu: 512,
                memory: 1024,
                essential: true,
                portMappings: [
                  {
                    containerPort: 8080,
                    protocol: 'tcp',
                  },
                ],
                environment: [
                  { name: 'SERVICE_NAME', value: 'worker-service' },
                  { name: 'ENVIRONMENT', value: environmentSuffix },
                ],
                secrets: [
                  { name: 'DB_CREDENTIALS', valueFrom: dbArn },
                  { name: 'API_KEYS', valueFrom: apiArn },
                ],
                logConfiguration: {
                  logDriver: 'awslogs',
                  options: {
                    'awslogs-group': logGroupName,
                    'awslogs-region': region,
                    'awslogs-stream-prefix': 'ecs',
                  },
                },
              },
            ])
          ),
        tags: {
          Name: `worker-task-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const schedulerTaskDefinition = new aws.ecs.TaskDefinition(
      `scheduler-task-${environmentSuffix}`,
      {
        family: `scheduler-service-${environmentSuffix}`,
        networkMode: 'awsvpc',
        requiresCompatibilities: ['FARGATE'],
        cpu: '512',
        memory: '1024',
        executionRoleArn: taskExecutionRole.arn,
        taskRoleArn: taskRole.arn,
        containerDefinitions: pulumi
          .all([schedulerEcrUrl, dbSecretArn, apiKeySecretArn, schedulerLogGroup.name, currentRegion.name])
          .apply(([ecrUrl, dbArn, apiArn, logGroupName, region]) =>
            JSON.stringify([
              {
                name: 'scheduler-service',
                image: `${ecrUrl}:latest`,
                cpu: 512,
                memory: 1024,
                essential: true,
                portMappings: [
                  {
                    containerPort: 8080,
                    protocol: 'tcp',
                  },
                ],
                environment: [
                  { name: 'SERVICE_NAME', value: 'scheduler-service' },
                  { name: 'ENVIRONMENT', value: environmentSuffix },
                ],
                secrets: [
                  { name: 'DB_CREDENTIALS', valueFrom: dbArn },
                  { name: 'API_KEYS', valueFrom: apiArn },
                ],
                logConfiguration: {
                  logDriver: 'awslogs',
                  options: {
                    'awslogs-group': logGroupName,
                    'awslogs-region': region,
                    'awslogs-stream-prefix': 'ecs',
                  },
                },
              },
            ])
          ),
        tags: {
          Name: `scheduler-task-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create ECS services
    const apiService = new aws.ecs.Service(
      `api-service-${environmentSuffix}`,
      {
        name: `api-service-${environmentSuffix}`,
        cluster: cluster.arn,
        taskDefinition: apiTaskDefinition.arn,
        desiredCount: 2,
        launchType: 'FARGATE',
        networkConfiguration: {
          subnets: privateSubnetIds,
          securityGroups: [ecsSecurityGroup.id],
          assignPublicIp: false,
        },
        loadBalancers: [
          {
            targetGroupArn: apiTargetGroup.arn,
            containerName: 'api-service',
            containerPort: 8080,
          },
        ],
        healthCheckGracePeriodSeconds: 60,
        tags: {
          Name: `api-service-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this, dependsOn: [listener] }
    );

    const workerService = new aws.ecs.Service(
      `worker-service-${environmentSuffix}`,
      {
        name: `worker-service-${environmentSuffix}`,
        cluster: cluster.arn,
        taskDefinition: workerTaskDefinition.arn,
        desiredCount: 2,
        launchType: 'FARGATE',
        networkConfiguration: {
          subnets: privateSubnetIds,
          securityGroups: [ecsSecurityGroup.id],
          assignPublicIp: false,
        },
        loadBalancers: [
          {
            targetGroupArn: workerTargetGroup.arn,
            containerName: 'worker-service',
            containerPort: 8080,
          },
        ],
        healthCheckGracePeriodSeconds: 60,
        tags: {
          Name: `worker-service-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this, dependsOn: [listener] }
    );

    const schedulerService = new aws.ecs.Service(
      `scheduler-service-${environmentSuffix}`,
      {
        name: `scheduler-service-${environmentSuffix}`,
        cluster: cluster.arn,
        taskDefinition: schedulerTaskDefinition.arn,
        desiredCount: 2,
        launchType: 'FARGATE',
        networkConfiguration: {
          subnets: privateSubnetIds,
          securityGroups: [ecsSecurityGroup.id],
          assignPublicIp: false,
        },
        loadBalancers: [
          {
            targetGroupArn: schedulerTargetGroup.arn,
            containerName: 'scheduler-service',
            containerPort: 8080,
          },
        ],
        healthCheckGracePeriodSeconds: 60,
        tags: {
          Name: `scheduler-service-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this, dependsOn: [listener] }
    );

    // Create auto-scaling targets
    const apiAutoScalingTarget = new aws.appautoscaling.Target(
      `api-autoscaling-${environmentSuffix}`,
      {
        maxCapacity: 10,
        minCapacity: 2,
        resourceId: pulumi.interpolate`service/${cluster.name}/${apiService.name}`,
        scalableDimension: 'ecs:service:DesiredCount',
        serviceNamespace: 'ecs',
      },
      { parent: this }
    );

    const workerAutoScalingTarget = new aws.appautoscaling.Target(
      `worker-autoscaling-${environmentSuffix}`,
      {
        maxCapacity: 10,
        minCapacity: 2,
        resourceId: pulumi.interpolate`service/${cluster.name}/${workerService.name}`,
        scalableDimension: 'ecs:service:DesiredCount',
        serviceNamespace: 'ecs',
      },
      { parent: this }
    );

    const schedulerAutoScalingTarget = new aws.appautoscaling.Target(
      `scheduler-autoscaling-${environmentSuffix}`,
      {
        maxCapacity: 10,
        minCapacity: 2,
        resourceId: pulumi.interpolate`service/${cluster.name}/${schedulerService.name}`,
        scalableDimension: 'ecs:service:DesiredCount',
        serviceNamespace: 'ecs',
      },
      { parent: this }
    );

    // Create auto-scaling policies based on CPU utilization
    new aws.appautoscaling.Policy(
      `api-cpu-scaling-${environmentSuffix}`,
      {
        name: `api-cpu-scaling-${environmentSuffix}`,
        policyType: 'TargetTrackingScaling',
        resourceId: apiAutoScalingTarget.resourceId,
        scalableDimension: apiAutoScalingTarget.scalableDimension,
        serviceNamespace: apiAutoScalingTarget.serviceNamespace,
        targetTrackingScalingPolicyConfiguration: {
          targetValue: 70,
          predefinedMetricSpecification: {
            predefinedMetricType: 'ECSServiceAverageCPUUtilization',
          },
          scaleInCooldown: 300,
          scaleOutCooldown: 60,
        },
      },
      { parent: this }
    );

    new aws.appautoscaling.Policy(
      `worker-cpu-scaling-${environmentSuffix}`,
      {
        name: `worker-cpu-scaling-${environmentSuffix}`,
        policyType: 'TargetTrackingScaling',
        resourceId: workerAutoScalingTarget.resourceId,
        scalableDimension: workerAutoScalingTarget.scalableDimension,
        serviceNamespace: workerAutoScalingTarget.serviceNamespace,
        targetTrackingScalingPolicyConfiguration: {
          targetValue: 70,
          predefinedMetricSpecification: {
            predefinedMetricType: 'ECSServiceAverageCPUUtilization',
          },
          scaleInCooldown: 300,
          scaleOutCooldown: 60,
        },
      },
      { parent: this }
    );

    new aws.appautoscaling.Policy(
      `scheduler-cpu-scaling-${environmentSuffix}`,
      {
        name: `scheduler-cpu-scaling-${environmentSuffix}`,
        policyType: 'TargetTrackingScaling',
        resourceId: schedulerAutoScalingTarget.resourceId,
        scalableDimension: schedulerAutoScalingTarget.scalableDimension,
        serviceNamespace: schedulerAutoScalingTarget.serviceNamespace,
        targetTrackingScalingPolicyConfiguration: {
          targetValue: 70,
          predefinedMetricSpecification: {
            predefinedMetricType: 'ECSServiceAverageCPUUtilization',
          },
          scaleInCooldown: 300,
          scaleOutCooldown: 60,
        },
      },
      { parent: this }
    );

    // Expose outputs
    this.albDnsName = alb.dnsName;
    this.clusterName = cluster.name;

    this.registerOutputs({
      albDnsName: this.albDnsName,
      clusterName: this.clusterName,
    });
  }
}
